File: 1.0.5/FileDownloader.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;
// Note: Regex imports are no longer needed
// import java.util.regex.Matcher;
// import java.util.regex.Pattern;

public class FileDownloader {

    // A set of the *only* folder names we are allowed to download from.
    private static final Set<String> ALLOWED_FOLDERS = Set.of(
        "files", "json", "others", "json_search"
    );

    // A single scanner to be used for all user input
    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        System.out.print("Enter the full URL to the text file (e.g., http://example.com/files.txt): ");
        String urlString = scanner.nextLine();

        URL textFileUrl;
        try {
            textFileUrl = new URL(urlString);
        } catch (MalformedURLException e) {
            System.err.println("Invalid URL provided. Please include http:// or https://");
            return;
        }

        try {
            // 1. Fetch the text content from the URL
            System.out.println("Fetching content from: " + textFileUrl);
            String textContent = fetchTextContent(textFileUrl);
            
            if (textContent == null || textContent.isEmpty()) {
                System.err.println("Could not fetch content or file is empty from " + textFileUrl);
            } else {
                System.out.println("Text file content fetched successfully.");
                // 2. Parse the text file line by line and download
                processFileLines(textContent, textFileUrl);
            }

        } catch (IOException e) {
            System.err.println("An error occurred: " + e.getMessage());
        }

        // Wait for user input before exiting
        System.out.println("\n--- All processing finished ---");
        System.out.println("Press Enter to exit...");
        scanner.nextLine(); // Wait for user to press Enter
        scanner.close(); // Close the scanner at the very end
    }

    /**
     * Downloads the text content of a given URL as a String.
     */
    private static String fetchTextContent(URL url) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("GET");
        connection.setRequestProperty("User-Agent", "Mozilla/5.0");
        connection.setInstanceFollowRedirects(true); 

        int responseCode = connection.getResponseCode();
        if (responseCode == HttpURLConnection.HTTP_OK) { // 200
            StringBuilder content = new StringBuilder();
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(
                    connection.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    content.append(line).append(System.lineSeparator());
                }
            }
            return content.toString();
        } else {
            System.err.println("GET request failed. Server responded with: " + responseCode);
            return null;
        }
    }

    /**
     * Parses the text content line by line and attempts to download
     * based on the specified folder structure rules.
     */
    private static void processFileLines(String fileContent, URL textFileUrl) {
        // \R is a regex for any universal newline character (e.g., \n, \r\n)
        String[] lines = fileContent.split("\\R");
        
        int totalLines = 0;
        int downloadedCount = 0;

        for (String line : lines) {
            totalLines++;
            String path = line.trim();

            // Skip empty lines or comment lines
            if (path.isEmpty() || path.startsWith("#")) {
                continue;
            }

            String[] parts = path.split("/");

            try {
                // --- This is the core logic from your request ---
                // We only want paths with exactly two parts: "folder/file.ext"
                if (parts.length == 2) {
                    String folderName = parts[0];
                    String fileName = parts[1];

                    if (fileName.isEmpty()) {
                        System.out.println("Skipping (directory path): " + path);
                    } else if (ALLOWED_FOLDERS.contains(folderName)) {
                        // This path matches all criteria.
                        // Resolve the relative path (e.g., "json/test.json")
                        // against the text file's URL (e.g., "http://example.com/files.txt")
                        // to get the full URL (e.g., "http://example.com/json/test.json")
                        URL absoluteUrl = new URL(textFileUrl, path);

                        // Attempt to download the file
                        if (downloadFile(absoluteUrl, folderName, fileName)) {
                            downloadedCount++;
                        }
                    } else {
                        System.out.println("Skipping (disallowed folder): " + path);
                    }
                } else if (parts.length == 1 && !parts[0].isEmpty()) {
                    System.out.println("Skipping (root file): " + path);
                } else if (parts.length > 2) {
                    System.out.println("Skipping (nested path): " + path);
                }
                // --- End of core logic ---

            } catch (MalformedURLException e) {
                System.err.println("Skipping malformed URL for path: " + path);
            }
        }
        
        // Print the final summary
        System.out.println("\n--- Summary ---");
        System.out.println("Total lines processed: " + totalLines);
        System.out.println("Files successfully downloaded: " + downloadedCount);
    }

    /**
     * Downloads a file.
     * @return true if download was successful, false otherwise.
     */
    private static boolean downloadFile(URL fileUrl, String folderName, String fileName) {
        File directory = new File(folderName);
        if (!directory.exists()) {
            if (directory.mkdir()) {
                System.out.println("Created directory: " + folderName);
            } else {
                System.err.println("Failed to create directory: " + folderName);
                return false;
            }
        }

        File localFile = new File(directory, fileName);

        if (localFile.exists()) {
            System.out.println("Skipping (already exists): " + localFile.getPath());
            return false; // Not a new download
        }

        try {
            HttpURLConnection fileConnection = (HttpURLConnection) fileUrl.openConnection();
            fileConnection.setRequestProperty("User-Agent", "Mozilla/5.0");
            fileConnection.setInstanceFollowRedirects(true);

            if (fileConnection.getResponseCode() == HttpURLConnection.HTTP_OK) {
                System.out.println("Downloading: " + fileUrl + " -> " + localFile.getPath());

                try (InputStream in = fileConnection.getInputStream();
                     FileOutputStream out = new FileOutputStream(localFile)) {
                    
                    byte[] buffer = new byte[4096];
                    int bytesRead;
                    while ((bytesRead = in.read(buffer)) != -1) {
                        out.write(buffer, 0, bytesRead);
                    }
                }
                System.out.println("Success: " + localFile.getPath());
                return true; // Download was successful
            } else {
                // This is common, e.g., 404 Not Found
                System.err.println("Failed to download " + fileUrl + ". Server responded: " 
                    + fileConnection.getResponseCode());
                return false;
            }
        } catch (IOException e) {
            System.err.println("Error downloading " + fileUrl + ": " + e.getMessage());
            return false;
        }
    }
}

File: 1.0.6.2/file_downloader.php
<?php
// A set of the *only* folder names we are allowed to download from.
define('ALLOWED_FOLDERS', ['files', 'json', 'others', 'json_search']);

if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $urlString = $_POST['url'] ?? '';
    
    if (empty($urlString)) {
        $error = "Please enter a URL";
    } else {
        // Validate URL format
        if (!filter_var($urlString, FILTER_VALIDATE_URL)) {
            $error = "Invalid URL provided. Please include http:// or https://";
        } else {
            // Process the download
            processDownload($urlString);
        }
    }
}

function processDownload($urlString) {
    echo "<h3>Processing URL: " . htmlspecialchars($urlString) . "</h3>";
    
    // 1. Fetch the text content from the URL
    $textContent = fetchTextContent($urlString);
    
    if ($textContent === null || empty($textContent)) {
        echo "<p style='color: red;'>Could not fetch content or file is empty from " . htmlspecialchars($urlString) . "</p>";
        return;
    }
    
    echo "<p>Text file content fetched successfully.</p>";
    
    // 2. Parse the text file line by line and download
    processFileLines($textContent, $urlString);
    
    echo "<p><strong>--- All processing finished ---</strong></p>";
}

function fetchTextContent($url) {
    $context = stream_context_create([
        'http' => [
            'method' => 'GET',
            'header' => 'User-Agent: Mozilla/5.0\r\n',
            'follow_location' => 1,
            'max_redirects' => 20
        ]
    ]);
    
    $content = @file_get_contents($url, false, $context);
    
    if ($content === false) {
        $http_response_header = $http_response_header ?? [];
        $responseCode = extractResponseCode($http_response_header);
        echo "<p style='color: red;'>GET request failed. Server responded with: " . $responseCode . "</p>";
        return null;
    }
    
    return $content;
}

function extractResponseCode($headers) {
    if (!empty($headers[0])) {
        preg_match('/HTTP\/\d\.\d\s+(\d+)/', $headers[0], $matches);
        return $matches[1] ?? 'Unknown';
    }
    return 'Unknown';
}

function processFileLines($fileContent, $textFileUrl) {
    $lines = explode("\n", str_replace("\r", "", $fileContent));
    $totalLines = 0;
    $downloadedCount = 0;
    
    foreach ($lines as $line) {
        $totalLines++;
        $path = trim($line);
        
        // Skip empty lines or comment lines
        if (empty($path) || strpos($path, '#') === 0) {
            continue;
        }
        
        $parts = explode('/', $path);
        
        // Core logic from Java version
        if (count($parts) == 2) {
            $folderName = $parts[0];
            $fileName = $parts[1];
            
            if (empty($fileName)) {
                echo "<p>Skipping (directory path): " . htmlspecialchars($path) . "</p>";
            } else if (in_array($folderName, ALLOWED_FOLDERS)) {
                // Build absolute URL
                $baseUrl = dirname($textFileUrl) . '/';
                $absoluteUrl = $baseUrl . $path;
                
                // Attempt to download the file
                if (downloadFile($absoluteUrl, $folderName, $fileName)) {
                    $downloadedCount++;
                }
            } else {
                echo "<p>Skipping (disallowed folder): " . htmlspecialchars($path) . "</p>";
            }
        } else if (count($parts) == 1 && !empty($parts[0])) {
            echo "<p>Skipping (root file): " . htmlspecialchars($path) . "</p>";
        } else if (count($parts) > 2) {
            echo "<p>Skipping (nested path): " . htmlspecialchars($path) . "</p>";
        }
    }
    
    // Print the final summary
    echo "<h4>--- Summary ---</h4>";
    echo "<p>Total lines processed: " . $totalLines . "</p>";
    echo "<p>Files successfully downloaded: " . $downloadedCount . "</p>";
}

function downloadFile($fileUrl, $folderName, $fileName) {
    // Create directory if it doesn't exist
    if (!is_dir($folderName)) {
        if (!mkdir($folderName, 0755, true)) {
            echo "<p style='color: red;'>Failed to create directory: " . htmlspecialchars($folderName) . "</p>";
            return false;
        }
    }
    
    $localFilePath = $folderName . '/' . $fileName;
    
    if (file_exists($localFilePath)) {
        echo "<p>Skipping (already exists): " . htmlspecialchars($localFilePath) . "</p>";
        return false;
    }
    
    $context = stream_context_create([
        'http' => [
            'method' => 'GET',
            'header' => 'User-Agent: Mozilla/5.0\r\n',
            'follow_location' => 1,
            'max_redirects' => 20
        ]
    ]);
    
    $fileContent = @file_get_contents($fileUrl, false, $context);
    
    if ($fileContent === false) {
        $http_response_header = $http_response_header ?? [];
        $responseCode = extractResponseCode($http_response_header);
        echo "<p style='color: red;'>Failed to download " . htmlspecialchars($fileUrl) . ". Server responded: " . $responseCode . "</p>";
        return false;
    }
    
    if (file_put_contents($localFilePath, $fileContent) !== false) {
        echo "<p>Success: " . htmlspecialchars($localFilePath) . "</p>";
        return true;
    } else {
        echo "<p style='color: red;'>Error writing file: " . htmlspecialchars($localFilePath) . "</p>";
        return false;
    }
}
?>

<!DOCTYPE html>
<html>
<head>
    <title>File Downloader</title>
</head>
<body>
    <h2>File Downloader</h2>
    
    <?php if (isset($error)): ?>
        <p style="color: red;"><?php echo htmlspecialchars($error); ?></p>
    <?php endif; ?>
    
    <form method="POST">
        <label for="url">Enter the full URL to the text file (e.g., http://example.com/files.txt):</label><br>
        <input type="text" id="url" name="url" style="width: 400px;" 
               value="<?php echo isset($_POST['url']) ? htmlspecialchars($_POST['url']) : ''; ?>">
        <br><br>
        <input type="submit" value="Download Files">
    </form>
</body>
</html>